src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -361,134 +369,266 @@ public class SqlMutationEngine : IMutationEngine
+                            }
+                        }
+
+                        string primaryKeyRouteForLocationHeader = isReadPermissionConfiguredForRole ? SqlResponseHelpers.ConstructPrimaryKeyRoute(context, mutationResultRow!.Columns, sqlMetadataProvider)
+                                                                                                    : string.Empty;
+
+                        if (context.OperationType is EntityActionOperation.Insert)
+                        {
+                            // Location Header is made up of the Base URL of the request and the primary key of the item created.
+                            // For POST requests, the primary key info would not be available in the URL and needs to be appended. So, the primary key of the newly created item
+                            // which is stored in the primaryKeyRoute is used to construct the Location Header.
+                            return SqlResponseHelpers.ConstructCreatedResultResponse(mutationResultRow!.Columns, selectOperationResponse, primaryKeyRouteForLocationHeader, isReadPermissionConfiguredForRole, isDatabasePolicyDefinedForReadAction, context.OperationType, GetBaseRouteFromConfig(_runtimeConfigProvider.GetConfig()), GetHttpContext());
+                        }
+
+                        if (context.OperationType is EntityActionOperation.Update || context.OperationType is EntityActionOperation.UpdateIncremental)
+                        {
+                            return SqlResponseHelpers.ConstructOkMutationResponse(mutationResultRow!.Columns, selectOperationResponse, isReadPermissionConfiguredForRole, isDatabasePolicyDefinedForReadAction);
+                        }
                     }
 
-                    string primaryKeyRoute = ConstructPrimaryKeyRoute(context, mutationResultRow.Columns, sqlMetadataProvider);
-                    // location will be updated in rest controller where httpcontext is available
-                    return new CreatedResult(location: primaryKeyRoute, OkMutationResponse(mutationResultRow.Columns).Value);
                 }
-
-                if (context.OperationType is EntityActionOperation.Update || context.OperationType is EntityActionOperation.UpdateIncremental)
+                finally
                 {
-                    // Nothing to update means we throw Exception
-                    if (mutationResultRow is null || mutationResultRow.Columns.Count == 0)
+                    if (selectOperationResponse is not null)
                     {
-                        throw new DataApiBuilderException(message: "No Update could be performed, record not found",
-                                                           statusCode: HttpStatusCode.PreconditionFailed,
-                                                           subStatusCode: DataApiBuilderException.SubStatusCodes.DatabaseOperationFailed);
+                        selectOperationResponse.Dispose();
                     }
```

Comment:
The instance of `JsonDocument` you use is declared outside the try/catch block this is finalizing. I added another comment where it would be more appropriate to add a `using` statement. The `using` statement also enables you to avoid using `finally` because finally is baked into using.

Reply:
In the docs, there are examples where the instance is declared outside the try-finally block but with the finally block containing the logic for disposing the instance. [Docs](https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/exceptions/how-to-execute-cleanup-code-using-finally)